Add notes to THREADS-03-04.md (#1772)
WebAssembly/meetings: threads/2025/THREADS-03-04.md

SYG (chat): it has to be an ephemeron map or it leaks, but maybe it leaks in jvm!

?? (chat): I’ve seen a number of comments cautioning folks from using Java TLS, so I think it leaks.

CW (chat): Technically it sort of doesn’t leak if you use real Java threads. But it leaks if you use thread pools and don’t manually clear. So everyone just says “manually clear all the time”, just in case the code is used in a thread pool.

RH: …do we need to compile to a different kind of shadow stack pointer?

TL: A task library would need some hand-written assembly to manage this… a pointer into a table of continuation references plus a pointer into linear memory to know where the shadow stack goes.

RH: Makes sense to me; my concern is that in resuming a stack, if the engine had spilled any thread local variables on the stack, it needs to unwind the stack and rewrite these values for the new stack. But then if the user written code clobbers it after that, all that work was wasted.

TL: The compiler would have to understand all this… It would need to mark the functions as "shared-suspendable" and would need to know not to stash the stack pointer in those functions. Seems complicated.

> AR: Circling back, I’m wondering how this is all related: when instantiating modules dynamically, wouldn’t the engine have to implement this under the hood. Don’t we need the same mechanism already for first-class thread locals? Why not expose this directly?

Reviewer: ```suggestion
AR: Circling back, I’m wondering how this is all related: when instantiating modules dynamically, wouldn’t the engine have to implement the same hashmap under the hood that languages with first-class thread locals need as well? Why not expose this directly?
```

TL: Thinking about LLVM, it understands the various TLS schemes (e.g., local_exec), because that’s all implemented in LLVM userspace, there is no need from those languages to provide this in WebAssembly. You’re right that the engine already does this for instances added to the store, I haven’t thought much about whether we need to expose it.

(TL presenting “Implementation Strategies”)

> AR: We already have global instances so this would be a pair of the global and instance…

Reviewer: ```suggestion
AR: We already have global instances independent of modules, so this would just be a pair of thread and global instance…
```
